File: mac-bridge/src/security/PathGuard.ts
import { realpath } from "node:fs/promises";
import { isAbsolute, relative } from "node:path";

export class PathGuard {
  private resolvedRoots?: Promise<readonly string[]>;

  constructor(private readonly roots: readonly string[]) {}

  async resolveAllowed(input: string): Promise<string> {
    if (!isAbsolute(input)) throw new Error("Media paths must be absolute");
    const resolved = await realpath(input);
    const roots = await (this.resolvedRoots ??= Promise.all(this.roots.map((root) => realpath(root))));
    const allowed = roots.some((root) => {
      const segment = relative(root, resolved);
      return segment === "" || (!segment.startsWith("..") && !isAbsolute(segment));
    });
    if (!allowed) throw new Error(`Media path is outside COMMANDER_MEDIA_ROOTS: ${resolved}`);
    return resolved;
  }
}

File: mac-bridge-go/internal/security/pathguard.go
package security

import (
	"errors"
	"fmt"
	"path/filepath"
	"strings"
)

type PathGuard struct {
	roots []string
}

func NewPathGuard(roots []string) *PathGuard {
	return &PathGuard{roots: append([]string(nil), roots...)}
}

func (g *PathGuard) ResolveAllowed(input string) (string, error) {
	if !filepath.IsAbs(input) {
		return "", errors.New("media paths must be absolute")
	}
	resolved, err := filepath.EvalSymlinks(input)
	if err != nil {
		return "", err
	}
	for _, root := range g.roots {
		canonicalRoot, rootErr := filepath.EvalSymlinks(root)
		if rootErr != nil {
			continue
		}
		relative, relErr := filepath.Rel(canonicalRoot, resolved)
		if relErr != nil {
			continue
		}
		if relative == "." || (relative != ".." && !strings.HasPrefix(relative, ".."+string(filepath.Separator)) && !filepath.IsAbs(relative)) {
			return resolved, nil
		}
	}
	return "", fmt.Errorf("media path is outside COMMANDER_MEDIA_ROOTS: %s", resolved)
}
